components/kyma-environment-broker/internal/hyperscaler/account_pool.go

```diff
@@ -0,0 +1,115 @@
+package hyperscaler
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type HyperscalerType string
+
+const (
+	GCP   HyperscalerType = "gcp"
+	Azure HyperscalerType = "azure"
+	AWS   HyperscalerType = "aws"
+)
+
+func HyperscalerTypeFromProviderString(provider string) (HyperscalerType, error) {
+
+	hyperscalerType := HyperscalerType(strings.ToLower(provider))
+
+	switch hyperscalerType {
+	case GCP, Azure, AWS:
+		return hyperscalerType, nil
+	}
+	return "", errors.Errorf("Unknown Hyperscaler provider type: %s", provider)
+}
+
+type Credentials struct {
+	CredentialName  string
+	HyperscalerType HyperscalerType
+	TenantName      string
+	CredentialData  map[string][]byte
+}
+
+type AccountPool interface {
+	Credentials(hyperscalerType HyperscalerType, tenantName string) (Credentials, error)
+}
+
+func NewAccountPool(secretsClient corev1.SecretInterface) AccountPool {
+	return &secretsAccountPool{
+		secretsClient: secretsClient,
+	}
+}
+
+type secretsAccountPool struct {
+	secretsClient corev1.SecretInterface
+	mux           sync.Mutex
+}
+
+func (p *secretsAccountPool) Credentials(hyperscalerType HyperscalerType, tenantName string) (Credentials, error) {
+
+	labelSelector := fmt.Sprintf("tenantName=%s,hyperscalerType=%s", tenantName, hyperscalerType)
+	secret, err := getSecret(p.secretsClient, labelSelector)
+
+	if err != nil {
+		return Credentials{}, err
+	}
+	if secret != nil {
+		return credentialsFromSecret(secret, hyperscalerType, tenantName), nil
+	}
+
+	labelSelector = fmt.Sprintf("!tenantName, hyperscalerType=%s", hyperscalerType)
+	// lock so that only one thread can fetch an unassigned secret and assign it (update secret with tenantName)
```

Comment:
the K8s provides optimistic locking, why don't we use it?

Reply:
This code has been already reviewed. This is correct